### ISSUE TITLE
setting PYTHONPATH to the stage directory

### ISSUE THREAD

**Issue by mwkrentel** [spack/spack#7231]
It looks like there is no way to export PYTHONPATH to a directory
inside the stage directory.  I think there's a timing issue between
setup_environment and the stage dir.

I'm writing a spack recipe for XED (Intel's x86 Encoder-Decoder).
XED uses its own python-based build system (mbuild) that lives in a
separate repository.

I can use a spack resource to clone the mbuild repo, but that kinda
puts the mbuild directory in the wrong place.  XED expects mbuild, by
default, to be in a side-by-side (sibling) directory.

Q1 -- Is it possible to put a spack resource somewhere other than as a
subdir of the main directory, in this case side-by-side?  I tried
resource with destination='../mbuild', but spack barfs at that.
Or, could I symlink it?

Anyway, XED will work with mbuild in any location, as long as you set
PYTHONPATH to the right directory.  I tried setup_environment() to set
PYTHONPATH.  But here, PYTHONPATH needs to be a subdir of the stage
dir and the stage dir is not known during setup_environment.  Inside
setup_environment, self.stage.source_path is None.

Q2 -- Is there a way to identify the stage dir in setup_environment?
I don't need the directory to exist, I just need to know the right
path name.

Q3 -- Or, is there somewhere later, after the stage dir is created
when I can set PYTHONPATH and it would be used in install?

I tried os.putenv('PYTHONPATH', ...) inside install, but that didn't
take effect.  I guess python had already read its environ variables
and didn't notice the change.  Is there a way to tell python to reread
its environment variables?

If absolutely necessary, I could make mbuild a separate, prereq
package.  The I could set PYTHONPATH to the mbuild prefix and it would
be known during setup_environment.

But that's overkill and what I really want is to set PYTHONPATH to
something in the stage directory.  It just looks like the spack timing
doesn't give me a way to do that.

Thanks,

--Mark

**Comment by scheibelp** [comment]
> Q3 -- Or, is there somewhere later, after the stage dir is created when I can set PYTHONPATH and it would be used in install?

Can you update `os.environ` from the `install` method?

**Comment by mwkrentel** [comment]
I've tried setting os.environ['PYTHONPATH'] inside install(), but it
doesn't seem to take effect.  I suspect the problem is that python has
already read its environ variables before install.  So, I'd need some
way to tell python to reread its environ vars.

I think the timing and launching of the python processes is a bit
complicated (I don't understand it).  package.py is, of course, a
python script.  Inside the install() method, I run:

   mfile = Executable('./mfile.py')

where mfile.py is the XED build script, also a python script.  And no,
it doesn't help to set os.environ['PYTHONPATH'] before Executable(...)
(tried that).

Something in mfile(...)  has to create another python process, but the
strange thing is that it doesn't seem to get the PYTHONPATH that I've
just set.

I suspect that spack is setting the environment for mfile(...) from
the values from setup_environment() and ignoring any changes after
that.  Why else would you need setup_environment?

So, I think I'm the victim of spack timing (again). :-) The logical
solution should be to use setup_environment() to set PYTHONPATH to
<stage-dir>/mbuild, but the stage-dir hasn't be set yet.  :-(
And the directory doesn't have to exist, I just have to know the
correct path.

**Comment by scheibelp** [comment]
`self.stage.source_path` should give you the path to the source. (and come to think of it, it is preferable to modify the environment in `setup_environment`)

> Why else would you need setup_environment?

To be clear Spack doesn't clear or ignore environment variable settings after `setup_environment`. It is more convenient to use though because the `spack_env` object has convenience methods, and also because if there are environment variables your package needs set at run time it can modify `run_env` and e.g. generated module files will include those modifications. It's also beneficial because those settings are recorded in `spack-build.env` (whereas environment variable settings done in `install` are not).

So in short, setting/appending `os.environ['PYTHONPATH']` in `install` *should* work.

As a side note:

> Inside the install() method, I run:
> mfile = Executable('./mfile.py')

I assume there is a shebang line there - what does it look like? It may be finding the non-Spack python. Assuming you have added a dependency on python you could have better luck doing

`python('mfile.py')`

Which will execute the `mfile.py` script with the Spack-built python.

**Comment by mwkrentel** [comment]
Well, this is interesting.  About two weeks ago, I submitted PR #7106
to add a new package for Intel XED.  It builds cleanly with Python
2.7.14 but fails with 3.6.3.  The fix is to set PYTHONPATH as I've
described.

My original PR was forked off commit f27c5e74ed on 1/28.  I tried the
os.environ['PYTHONPATH'] fix in my PR branch, but it didn't work.  And
setup_environment also doesn't work because self.stage isn't set.

Today, following your suggestions, I updated to the latest spack and
retried the os.environ fix.  And now it works!

So, I guess somebody in the last two weeks fixed something that also
fixes my problem.  There's roughly 100 commits in that range.  I
haven't tried binary search, but there's one commit that stands out as
a likely possibility.

I think Massimiliano Culpo's commit b7baeb542d on 2/12 fixes the
problem.  The issue #7128 mentions the error message, "ImportError: No
module named site," which is the same error I got.

Anyway, I think I now have a clean path to victory.  I'll update my
spack fork, apply the fix and resubmit the PR.

Btw, I don't know if it matters, but if you try to use
self.stage.source_path from within setup_environment, you just get
'None'.  But I can use the os.environ fix.

Also, if you'd like to take a look at the PR for Intel XED, I'd
appreciate that.  XED is a library for cracking X86 instructions.
It's been sitting idle with a red X for codecov. :-(

Thanks!

**Comment by scheibelp** [comment]
Sounds like this issue is resolved. Please reopen if it comes up again.

> Also, if you'd like to take a look at the PR for Intel XED, I'd appreciate that. XED is a library for cracking X86 instructions.

Can you add a mention there? Thanks!
